### ISSUE TITLE
get custom extension property value in client credentials flow 

### ISSUE THREAD

**Issue by sven5** [azure-ad-b2c/samples#436]
Hi @yoelhor  - I hope you could help me.

Let's say I have a custom extensions property in the app registration's manifest:
`"extension_5b087af1548c4d01832ff07d1ec48150_CustomerNumber": "123456"`

How can I add this value to the token by using a custom policy?

**Comment by JasSuri** [comment]
You cannot read data from an App Registration object.
You can use AAD or REST API technical profiles to interact with data sources.
If this value is static, then you could hardcode it using defaultValue in custom policy.

**Comment by sven5** [comment]
@JasSuri yes, I'd like to use AAD technical profile to read data from the application registration.
Would you mov mind showing an example?

**Comment by JasSuri** [comment]
You cannot use AAD technical profile to read data from the application registration. Only from User objects.
An interface to directly pull information from an Application object doesn't exist within IEF.

>What should I do if I wanted to have the displayName property of the app in the token?

Only way to do that, would be to call your own API inside the policy, and have your API call MS Graph API and return the value to IEF.

**Comment by sven5** [comment]
Hi @JasSuri  .

Thanks for clarification. In parallel, I already contacted Azure support and they told me that it should be possible to read application data by just using AAD technical profile. Now I requested a sample from them. 
However, my initial assumption was right, that we have to implement a REST API for this particular case.

**Comment by JasSuri** [comment]
How did you do this:

> I have a custom extensions property in the app registration's manifest:
"extension_5b087af1548c4d01832ff07d1ec48150_CustomerNumber": "123456"

I was not even aware this was possible. As far i know, the manifest could be used to define/extend the schema, but not actually store some value for this schema.

**Comment by sven5** [comment]
At first, I defined the CustomerNumber as extension property for User and Application types. Then, 
I just edited the manifest json and added the property. It saved without error.

**Comment by BouvetMarius** [comment]
> For anyone still interested, I ended up by implementing a REST API using an Azure Function to gather some information about the application registration. This way, I'm able to read the value of my custom extension property by using [MS Graph SDK](https://github.com/microsoftgraph/msgraph-sdk-dotnet).
> 
> I was struggling to get the app's `objectId` from the client credential technical profile and it seems that this is impossible.
> 
> I successfully created a REST technical profile and retrieved the `clientId` using [claim resolvers](https://docs.microsoft.com/en-us/azure/active-directory-b2c/claim-resolver-overview).
> 
> ```
> <ClaimsProvider>
> 			<DisplayName>Get Customer Number</DisplayName>
> 			<TechnicalProfiles>
> 				
> 				<TechnicalProfile Id="REST-GetCustomerNumber">
> 					<DisplayName/>
> 					<Protocol Name="Proprietary" Handler="Web.TPEngine.Providers.RestfulProvider, Web.TPEngine, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null"/>
> 					<Metadata>
> 						<Item Key="ServiceUrl">https://my-azure-function.azurewebsites.net/api/GetCustomerNumberForApplication</Item>
> 						<Item Key="AuthenticationType">ApiKeyHeader</Item>
> 						<Item Key="SendClaimsIn">Body</Item>
> 						<Item Key="DebugMode">true</Item>
> 					</Metadata>
> 					<CryptographicKeys>
> 						<Key Id="x-functions-key" StorageReferenceId="B2C_1A_GetCustomerNumberKey"/>
> 					</CryptographicKeys>
> 					<InputClaims>
> 						<InputClaim ClaimTypeReferenceId="clientId" DefaultValue="{OIDC:ClientId}" AlwaysUseDefaultValue="true" />
> 						<InputClaim ClaimTypeReferenceId="tenantId" DefaultValue="{Policy:TenantObjectId}" AlwaysUseDefaultValue="true" />
> 					</InputClaims>
> 					<OutputClaims>
> 						<OutputClaim ClaimTypeReferenceId="customerNumber" PartnerClaimType="customerNumber" DefaultValue="[CUSTNO]" />
> 					</OutputClaims>
> 					<UseTechnicalProfileForSessionManagement ReferenceId="SM-Noop"/>
> 				</TechnicalProfile>
> 			</TechnicalProfiles>
> 		</ClaimsProvider>
> ```

Hi @sven5 

How do you get the ClientId ? Im trying to solve almost the same thing, and could need some help

**Comment by sven5** [comment]
@BouvetMarius  I already mentioned twice how to get the ClientId:

1.

> retrieved the clientId using [claim resolvers](https://docs.microsoft.com/en-us/azure/active-directory-b2c/claim-resolver-overview).
2. See the following line in my Technical Profile:
`<InputClaim ClaimTypeReferenceId="clientId" DefaultValue="{OIDC:ClientId}" AlwaysUseDefaultValue="true" />`
